File: next_app/next_app/report/sales_invoice/sales_invoice.py
import frappe

@frappe.whitelist()
def execute(filters=None):
    columns = get_columns()
    data, total = get_data(filters)
    summary = get_summary(filters)
    return columns, data, total, summary

def get_columns():
    return [
        {"label": "ID", "fieldname": "name", "fieldtype": "Link", "options": "Sales Invoice", "width": 150},
        {"label": "Customer", "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 150},
        {"label": "Posting Date", "fieldname": "posting_date", "fieldtype": "Date", "width": 120},
        {"label": "Posting Time", "fieldname": "posting_time", "fieldtype": "Time", "width": 120},
        {"label": "Grand Total", "fieldname": "grand_total", "fieldtype": "Currency", "width": 120},
        {"label": "Status", "fieldname": "status", "fieldtype": "Data", "width": 100},
        {"label": "Owner", "fieldname": "owner", "fieldtype": "Data", "width": 100},
        {"label": "Creation", "fieldname": "creation", "fieldtype": "Datetime", "width": 150},
        {"label": "Currency", "fieldname": "currency", "fieldtype": "Link", "options": "Currency", "width": 100},
        {"label": "Update Stock", "fieldname": "update_stock", "fieldtype": "Check", "width": 100},
        {"label": "Items", "fieldname": "items", "fieldtype": "Table", "width": 400}
    ]

def get_data(filters):
    conditions = get_conditions(filters)

    limit_start = filters.get("limit_start", 0)
    limit_page_length = filters.get("limit_page_length", 10)
    order_by = filters.get("order_by", "posting_date")
    order = filters.get("order", "desc")

    total = frappe.db.sql(f"""
        SELECT COUNT(name)
        FROM `tabSales Invoice`
        WHERE 1=1 {conditions}
    """)[0][0]

    invoices = frappe.db.sql(f"""
        SELECT
            name, customer, posting_date, posting_time, grand_total, status, owner, creation, currency, update_stock, outstanding_amount
        FROM
            `tabSales Invoice`
        WHERE
            1=1
            {conditions}
        ORDER BY
            {order_by} {order}
        LIMIT
            {limit_start}, {limit_page_length}
    """, as_dict=True)

    for invoice in invoices:
        items = frappe.db.sql(f"""
            SELECT
                item_code, item_name, qty, rate, amount, delivered_qty
            FROM
                `tabSales Invoice Item`
            WHERE
                parent = '{invoice.name}'
        """, as_dict=True)
        invoice["items"] = items

    return invoices, total

def get_summary(filters):
    conditions = get_conditions(filters)
    active_currencies = frappe.get_all("Currency", filters={"enabled": 1}, fields=["name"])
    currency_list = [c["name"] for c in active_currencies]
    
    summary_data = {}
    for currency in currency_list:
        summary_data[currency] = {
            "currency": currency,
            "total_sales": 0,
            "paid_sales": 0,
            "credit_sales": 0,
        }

    sales_data = frappe.db.sql(f"""
        SELECT
            currency,
            IFNULL(SUM(grand_total), 0) as total_sales,
            IFNULL(SUM(CASE WHEN status = 'Paid' THEN grand_total ELSE 0 END), 0) as paid_sales,
            IFNULL(SUM(CASE WHEN status != 'Paid' AND status != 'Cancelled' THEN outstanding_amount ELSE 0 END), 0) as credit_sales
        FROM
            `tabSales Invoice`
        WHERE docstatus = 1 {conditions}
        GROUP BY
            currency
    """, as_dict=True)

    for row in sales_data:
        if row.currency in summary_data:
            summary_data[row.currency] = row

    return list(summary_data.values())

def get_conditions(filters):
    conditions = ""
    if filters:
        if filters.get("customer"):
            conditions += f" AND customer LIKE '%{filters.get('customer')}%'"
        if filters.get("name"):
            conditions += f" AND name LIKE '%{filters.get('name')}%'"
        if filters.get("status"):
            conditions += f" AND status = '{filters.get('status')}'"
        if filters.get("date_range"):
            start_date, end_date = filters.get("date_range")
            conditions += f" AND posting_date BETWEEN '{start_date}' AND '{end_date}'"
        if filters.get("warehouse"):
            conditions += f" AND set_warehouse = '{filters.get('warehouse')}'"
    return conditions

File: next/src/types/Accounts/SalesInvoice.ts
import type { a } from "node_modules/@tanstack/react-query-devtools/build/modern/ReactQueryDevtools-Cn7cKi7o"
import type { SalesInvoiceItem } from "./SalesInvoiceItem"

export interface SalesInvoice{
	name: string
	creation: string
	modified: string
	owner: string
	modified_by: string
	docstatus: 0 | 1 | 2
	parent?: string
	parentfield?: string
	parenttype?: string
	idx?: number
	/**	Title : Data	*/
	title?: string
	/**	Series : Select	*/
	naming_series: "ACC-SINV-.YYYY.-" | "ACC-SINV-RET-.YYYY.-"
	/**	Customer : Link - Customer	*/
	customer?: string
	/**	Customer Name : Small Text	*/
	customer_name?: string
	/**	Tax Id : Data	*/
	tax_id?: string
	/**	Company : Link - Company	*/
	company: string
	/**	Company Tax ID : Data	*/
	company_tax_id?: string
	/**	Date : Date	*/
	posting_date: string
	/**	Posting Time : Time	*/
	posting_time?: string
	/**	Edit Posting Date and Time : Check	*/
	set_posting_time?: 0 | 1
	/**	Payment Due Date : Date	*/
	due_date?: string
	/**	Include Payment (POS) : Check	*/
	is_pos?: 0 | 1
	/**	POS Profile : Link - POS Profile	*/
	pos_profile?: string
	/**	Is Consolidated : Check	*/
	is_consolidated?: 0 | 1
	/**	Is Return (Credit Note) : Check	*/
	is_return?: 0 | 1
	/**	Return Against : Link - Sales Invoice	*/
	return_against?: string
	/**	Update Outstanding for Self : Check - Credit Note will update it's own outstanding amount, even if "Return Against" is specified.	*/
	update_outstanding_for_self?: 0 | 1
	/**	Update Billed Amount in Sales Order : Check	*/
	update_billed_amount_in_sales_order?: 0 | 1
	/**	Update Billed Amount in Delivery Note : Check	*/
	update_billed_amount_in_delivery_note?: 0 | 1
	/**	Is Rate Adjustment Entry (Debit Note) : Check - Issue a debit note with 0 qty against an existing Sales Invoice	*/
	is_debit_note?: 0 | 1
	/**	Amended From : Link - Sales Invoice	*/
	amended_from?: string
	/**	Cost Center : Link - Cost Center	*/
	cost_center?: string
	/**	Project : Link - Project	*/
	project?: string
	/**	Currency : Link - Currency	*/
	currency: string
	/**	Exchange Rate : Float - Rate at which Customer Currency is converted to customer's base currency	*/
	conversion_rate: number
	/**	Price List : Link - Price List	*/
	selling_price_list: string
	/**	Price List Currency : Link - Currency	*/
	price_list_currency: string
	/**	Price List Exchange Rate : Float - Rate at which Price list currency is converted to customer's base currency	*/
	plc_conversion_rate: number
	/**	Ignore Pricing Rule : Check	*/
	ignore_pricing_rule?: 0 | 1
	/**	Scan Barcode : Data	*/
	scan_barcode?: string
	/**	Update Stock : Check	*/
	update_stock?: 0 | 1
	/**	Source Warehouse : Link - Warehouse	*/
	set_warehouse?: string
	/**	Set Target Warehouse : Link - Warehouse	*/
	set_target_warehouse?: string
	/**	Items : Table - Sales Invoice Item	*/
	items: any
	/**	Total Quantity : Float	*/
	total_qty?: number
	/**	Total Net Weight : Float	*/
	total_net_weight?: number
	/**	Total (Company Currency) : Currency	*/
	base_total?: number
	/**	Net Total (Company Currency) : Currency	*/
	base_net_total: number
	/**	Total : Currency	*/
	total?: number
	/**	Net Total : Currency	*/
	net_total?: number
	/**	Tax Category : Link - Tax Category	*/
	tax_category?: string
	/**	Sales Taxes and Charges Template : Link - Sales Taxes and Charges Template	*/
	taxes_and_charges?: string
	/**	Shipping Rule : Link - Shipping Rule	*/
	shipping_rule?: string
	/**	Incoterm : Link - Incoterm	*/
	incoterm?: string
	/**	Named Place : Data	*/
	named_place?: string
	/**	Sales Taxes and Charges : Table - Sales Taxes and Charges	*/
	taxes?: any
	/**	Total Taxes and Charges (Company Currency) : Currency	*/
	base_total_taxes_and_charges?: number
	/**	Total Taxes and Charges : Currency	*/
	total_taxes_and_charges?: number
	/**	Grand Total (Company Currency) : Currency	*/
	base_grand_total: number
	/**	Rounding Adjustment (Company Currency) : Currency	*/
	base_rounding_adjustment?: number
	/**	Rounded Total (Company Currency) : Currency	*/
	base_rounded_total?: number
	/**	In Words (Company Currency) : Small Text - In Words will be visible once you save the Sales Invoice.	*/
	base_in_words?: string
	/**	Grand Total : Currency	*/
	grand_total: number
	/**	Rounding Adjustment : Currency	*/
	rounding_adjustment?: number
	/**	Use Company default Cost Center for Round off : Check	*/
	use_company_roundoff_cost_center?: 0 | 1
	/**	Rounded Total : Currency	*/
	rounded_total?: number
	/**	In Words : Small Text	*/
	in_words?: string
	/**	Total Advance : Currency	*/
	total_advance?: number
	/**	Outstanding Amount : Currency	*/
	outstanding_amount?: number
	/**	Disable Rounded Total : Check	*/
	disable_rounded_total?: 0 | 1
	/**	Apply Additional Discount On : Select	*/
	apply_discount_on?: "" | "Grand Total" | "Net Total"
	/**	Additional Discount Amount (Company Currency) : Currency	*/
	base_discount_amount?: number
	/**	Is Cash or Non Trade Discount : Check	*/
	is_cash_or_non_trade_discount?: 0 | 1
	/**	Discount Account : Link - Account	*/
	additional_discount_account?: string
	/**	Additional Discount Percentage : Float	*/
	additional_discount_percentage?: number
	/**	Additional Discount Amount : Currency	*/
	discount_amount?: number
	/**	Taxes and Charges Calculation : Text Editor	*/
	other_charges_calculation?: string
	/**	Pricing Rule Detail : Table - Pricing Rule Detail	*/
	pricing_rules?: any
	/**	Packed Items : Table - Packed Item	*/
	packed_items?: any
	/**	Time Sheets : Table - Sales Invoice Timesheet	*/
	timesheets?: any
	/**	Total Billing Hours : Float	*/
	total_billing_hours?: number
	/**	Total Billing Amount : Currency	*/
	total_billing_amount?: number
	/**	Cash/Bank Account : Link - Account	*/
	cash_bank_account?: string
	/**	Sales Invoice Payment : Table - Sales Invoice Payment	*/
	payments?: any
	/**	Paid Amount (Company Currency) : Currency	*/
	base_paid_amount?: number
	/**	Paid Amount : Currency	*/
	paid_amount?: number
	/**	Base Change Amount (Company Currency) : Currency	*/
	base_change_amount?: number
	/**	Change Amount : Currency	*/
	change_amount?: number
	/**	Account for Change Amount : Link - Account	*/
	account_for_change_amount?: string
	/**	Allocate Advances Automatically (FIFO) : Check	*/
	allocate_advances_automatically?: 0 | 1
	/**	Only Include Allocated Payments : Check - Advance payments allocated against orders will only be fetched	*/
	only_include_allocated_payments?: 0 | 1
	/**	Advances : Table - Sales Invoice Advance	*/
	advances?: any
	/**	Write Off Amount : Currency	*/
	write_off_amount?: number
	/**	Write Off Amount (Company Currency) : Currency	*/
	base_write_off_amount?: number
	/**	Write Off Outstanding Amount : Check	*/
	write_off_outstanding_amount_automatically?: 0 | 1
	/**	Write Off Account : Link - Account	*/
	write_off_account?: string
	/**	Write Off Cost Center : Link - Cost Center	*/
	write_off_cost_center?: string
	/**	Redeem Loyalty Points : Check	*/
	redeem_loyalty_points?: 0 | 1
	/**	Loyalty Points : Int	*/
	loyalty_points?: number
	/**	Loyalty Amount : Currency	*/
	loyalty_amount?: number
	/**	Loyalty Program : Link - Loyalty Program	*/
	loyalty_program?: string
	/**	Redemption Account : Link - Account	*/
	loyalty_redemption_account?: string
	/**	Redemption Cost Center : Link - Cost Center	*/
	loyalty_redemption_cost_center?: string
	/**	Customer Address : Link - Address	*/
	customer_address?: string
	/**	Address : Small Text	*/
	address_display?: string
	/**	Contact Person : Link - Contact	*/
	contact_person?: string
	/**	Contact : Small Text	*/
	contact_display?: string
	/**	Mobile No : Small Text	*/
	contact_mobile?: string
	/**	Contact Email : Data	*/
	contact_email?: string
	/**	Territory : Link - Territory	*/
	territory?: string
	/**	Shipping Address Name : Link - Address	*/
	shipping_address_name?: string
	/**	Shipping Address : Small Text	*/
	shipping_address?: string
	/**	Dispatch Address Name : Link - Address	*/
	dispatch_address_name?: string
	/**	Dispatch Address : Small Text	*/
	dispatch_address?: string
	/**	Company Address Name : Link - Address	*/
	company_address?: string
	/**	Company Address : Small Text	*/
	company_address_display?: string
	/**	Ignore Default Payment Terms Template : Check	*/
	ignore_default_payment_terms_template?: 0 | 1
	/**	Payment Terms Template : Link - Payment Terms Template	*/
	payment_terms_template?: string
	/**	Payment Schedule : Table - Payment Schedule	*/
	payment_schedule?: any
	/**	Terms : Link - Terms and Conditions	*/
	tc_name?: string
	/**	Terms and Conditions Details : Text Editor	*/
	terms?: string
	/**	Customer's Purchase Order : Data	*/
	po_no?: string
	/**	Customer's Purchase Order Date : Date	*/
	po_date?: string
	/**	Debit To : Link - Account	*/
	debit_to: string
	/**	Party Account Currency : Link - Currency	*/
	party_account_currency?: string
	/**	Is Opening Entry : Select	*/
	is_opening?: "No" | "Yes"
	/**	Unrealized Profit / Loss Account : Link - Account - Unrealized Profit / Loss account for intra-company transfers	*/
	unrealized_profit_loss_account?: string
	/**	Against Income Account : Small Text	*/
	against_income_account?: string
	/**	Sales Partner : Link - Sales Partner	*/
	sales_partner?: string
	/**	Amount Eligible for Commission : Currency	*/
	amount_eligible_for_commission?: number
	/**	Commission Rate (%) : Float	*/
	commission_rate?: number
	/**	Total Commission : Currency	*/
	total_commission?: number
	/**	Sales Contributions and Incentives : Table - Sales Team	*/
	sales_team?: any
	/**	Letter Head : Link - Letter Head	*/
	letter_head?: string
	/**	Group same items : Check	*/
	group_same_items?: 0 | 1
	/**	Print Heading : Link - Print Heading	*/
	select_print_heading?: string
	/**	Print Language : Data	*/
	language?: string
	/**	Subscription : Link - Subscription	*/
	subscription?: string
	/**	From Date : Date	*/
	from_date?: string
	/**	Auto Repeat : Link - Auto Repeat	*/
	auto_repeat?: string
	/**	To Date : Date	*/
	to_date?: string
	/**	Status : Select	*/
	status?: "" | "Draft" | "Return" | "Credit Note Issued" | "Submitted" | "Paid" | "Partly Paid" | "Unpaid" | "Unpaid and Discounted" | "Partly Paid and Discounted" | "Overdue and Discounted" | "Overdue" | "Cancelled" | "Internal Transfer"
	/**	Inter Company Invoice Reference : Link - Purchase Invoice	*/
	inter_company_invoice_reference?: string
	/**	Campaign : Link - Campaign	*/
	campaign?: string
	/**	Represents Company : Link - Company - Company which internal customer represents	*/
	represents_company?: string
	/**	Source : Link - Lead Source	*/
	source?: string
	/**	Customer Group : Link - Customer Group	*/
	customer_group?: string
	/**	Is Internal Customer : Check	*/
	is_internal_customer?: 0 | 1
	/**	Is Discounted : Check	*/
	is_discounted?: 0 | 1
	/**	Remarks : Small Text	*/
	remarks?: string
}